### PR TITLE
 builder: deneb and handling blobs

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**
 Feature

**What does this PR do? Why is it needed?**

Introduces deneb block and blobs to builder 

implements changes from
https://github.com/ethereum/builder-specs/pull/80
https://github.com/ethereum/builder-specs/pull/79
https://github.com/ethereum/builder-specs/pull/61

https://github.com/ethereum/consensus-specs/pull/3391
https://github.com/ethereum/consensus-specs/pull/3392

part of #12248

### CODE REVIEW

File: beacon-chain/builder/service.go [hunk 1]
```diff
@@ -87,18 +90,21 @@ func (s *Service) Stop() error {
...
 	defer span.End()
 	start := time.Now()
 	defer func() {
 		submitBlindedBlockLatency.Observe(float64(time.Since(start).Milliseconds()))
 	}()
 	if s.c == nil {
-		return nil, ErrNoBuilder
+		return nil, nil, ErrNoBuilder
+	}
+	if blobs != nil && uint64(len(blobs)) > fieldparams.MaxBlobsPerBlock {
```
Comment (james-prysm) [inline beacon-chain/builder/service.go:103]: Do i need this check here?

File: api/client/builder/client.go [hunk 2]
```diff
@@ -327,22 +342,58 @@ func (c *Client) SubmitBlindedBlock(ctx context.Context, sb interfaces.ReadOnlyS
...
+			return nil, nil, errors.Wrapf(err, "could not wrap execution payload to interface")
+		}
+		return payload, nil, nil
+	case version.Deneb:
+		psb, err := sb.PbBlindedDenebBlock()
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "could not get protobuf block")
+		}
+
+		b := &ethpb.SignedBlindedBeaconBlockAndBlobsDeneb{Block: psb, Blobs: blobs}
```
Comment (james-prysm) [inline api/client/builder/client.go:369]: any additional checks needed for blobs here?

File: api/client/builder/types.go [hunk 3]
```diff
@@ -1024,6 +1024,272 @@ func (b *BlindedBeaconBlockBodyCapella) MarshalJSON() ([]byte, error) {
...
+type BlobsBundle struct {
+	Commitments []hexutil.Bytes `json:"commitments"`
+	Proofs      []hexutil.Bytes `json:"proofs"`
+	Blobs       []hexutil.Bytes `json:"blobs"`
+}
+
+func (b BlobsBundle) ToProto() (*v1.BlobsBundle, error) {
+	commitments := make([][]byte, len(b.Commitments))
+	for i := range b.Commitments {
+		if len(b.Commitments[i]) != 48 {
```
Comment (terencechain) [inline api/client/builder/types.go:1215]: Replace 48 with `fieldparams.BLSPubkeyLength`

File: api/client/builder/types.go [hunk 4]
```diff
@@ -1024,6 +1025,278 @@ func (b *BlindedBeaconBlockBodyCapella) MarshalJSON() ([]byte, error) {
...
+	Commitments []hexutil.Bytes `json:"commitments"`
+	Proofs      []hexutil.Bytes `json:"proofs"`
+	Blobs       []hexutil.Bytes `json:"blobs"`
+}
+
+func (b BlobsBundle) ToProto() (*v1.BlobsBundle, error) {
+	commitments := make([][]byte, len(b.Commitments))
+	for i := range b.Commitments {
+		if len(b.Commitments[i]) != fieldparams.BLSPubkeyLength {
+			return nil, fmt.Errorf("commitments length %d is not %d", len(b.Commitments[i]), fieldparams.BLSPubkeyLength)
```
Comment (rkapka) [inline api/client/builder/types.go:1217]:
```suggestion
			return nil, fmt.Errorf("commitment length %d is not %d", len(b.Commitments[i]), fieldparams.BLSPubkeyLength)
```

File: api/client/builder/types.go [hunk 5]
```diff
@@ -1024,6 +1025,278 @@ func (b *BlindedBeaconBlockBodyCapella) MarshalJSON() ([]byte, error) {
...
+	BlobsBundle      *BlobsBundle           `json:"blobs_bundle"`
+}
+
+type BlobsBundle struct {
+	Commitments []hexutil.Bytes `json:"commitments"`
+	Proofs      []hexutil.Bytes `json:"proofs"`
+	Blobs       []hexutil.Bytes `json:"blobs"`
+}
+
+func (b BlobsBundle) ToProto() (*v1.BlobsBundle, error) {
```
Comment (rkapka) [inline api/client/builder/types.go:1213]: We make unsafe copies everywhere

File: api/client/builder/types.go [hunk 6]
```diff
@@ -1024,6 +1025,278 @@ func (b *BlindedBeaconBlockBodyCapella) MarshalJSON() ([]byte, error) {
...
+		return nil, nil, err
+	}
+	bundle, err := r.Data.BlobsBundle.ToProto()
+	if err != nil {
+		return nil, nil, err
+	}
+	return payload, bundle, nil
+}
+
+func (p *ExecutionPayloadDeneb) ToProto() (*v1.ExecutionPayloadDeneb, error) {
```
Comment (rkapka) [inline api/client/builder/types.go:1265]: We make unsafe copies everywhere

File: proto/engine/v1/execution_engine.proto [hunk 7]
```diff
@@ -218,6 +218,16 @@ message BlobsBundle {
 	repeated bytes blobs = 3 [(ethereum.eth.ext.ssz_size) = "?,131072", (ethereum.eth.ext.ssz_max) = "4"];
 }
 
+// BlindedBlobsBundle is retrieved through the builder-api from /eth/v1/builder/header/{slot}/{parent_hash}/{pubkey} after the Deneb hardfork.
+message BlindedBlobsBundle {
+	// The KZG commitments of the blobs.
+	repeated bytes kzg_commitments = 1 [(ethereum.eth.ext.ssz_size) = "?,48", (ethereum.eth.ext.ssz_max) = "4"];
+	// The proofs of the blobs.
+	repeated bytes proofs = 2 [(ethereum.eth.ext.ssz_size) = "?,48", (ethereum.eth.ext.ssz_max) = "4"];
+	// The blobs itself.
```
Comment (rkapka) [inline proto/engine/v1/execution_engine.proto:227]:
```suggestion
	// The blob roots.
```

File: testing/util/blob.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
+package util
+
+import (
+	fieldparams "github.com/prysmaticlabs/prysm/v4/config/fieldparams"
+	ethpb "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+)
+
+// HydrateSignedBlindedBlobSidecar hydrates a signed blinded blob sidecar with correct field length sizes
+// to comply with fssz marshalling and unmarshalling rules.
```
Comment (rkapka) [inline testing/util/blob.go:9]:
```suggestion
// to comply with SSZ marshalling and unmarshalling rules.
```

File: testing/util/blob.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+func HydrateSignedBlindedBlobSidecar(b *ethpb.SignedBlindedBlobSidecar) *ethpb.SignedBlindedBlobSidecar {
+	if b.Signature == nil {
+		b.Signature = make([]byte, fieldparams.BLSSignatureLength)
+	}
+	b.Message = HydrateBlindedBlobSidecar(b.Message)
+	return b
+}
+
+// HydrateBlindedBlobSidecar hydrates a blinded blob sidecar with correct field length sizes
+// to comply with fssz marshalling and unmarshalling rules.
```
Comment (rkapka) [inline testing/util/blob.go:19]:
```suggestion
// to comply with SSZ marshalling and unmarshalling rules.
```

File: testing/util/merge.go [hunk 10]
```diff
@@ -43,6 +43,10 @@ func NewBlindedBeaconBlockDeneb() *ethpb.SignedBlindedBeaconBlockDeneb {
 	return HydrateSignedBlindedBeaconBlockDeneb(&ethpb.SignedBlindedBeaconBlockDeneb{})
 }
 
+func NewBlindedBlobSidecar() *ethpb.SignedBlindedBlobSidecar {
```
Comment (rkapka) [inline testing/util/merge.go:46]: godoc missing

File: api/client/builder/types.go [hunk 11]
```diff
@@ -1024,6 +1025,278 @@ func (b *BlindedBeaconBlockBodyCapella) MarshalJSON() ([]byte, error) {
 	})
 }
 
+type ExecHeaderResponseDeneb struct {
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply api/client/builder/types.go:1028]: I'll add it for deneb here and address it in develop for old things

File: api/client/builder/types.go [hunk 12]
```diff
@@ -1024,6 +1025,278 @@ func (b *BlindedBeaconBlockBodyCapella) MarshalJSON() ([]byte, error) {
...
+		return err
+	}
+	ep := ExecutionPayloadHeaderDeneb(*uc)
+	*h = ep
+	var err error
+	h.ExecutionPayloadHeaderDeneb, err = h.ToProto()
+	return err
+}
+
+func (h *ExecutionPayloadHeaderDeneb) ToProto() (*v1.ExecutionPayloadHeaderDeneb, error) {
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply api/client/builder/types.go:1170]: added it for deneb regarding to proto function, will need to open a new one against develop for older ones.

File: api/client/builder/types.go (hunk 5 above)
Comment (james-prysm) [reply api/client/builder/types.go:1233]: added in

**Review by terencechain — DISMISSED** [review]
It's not entirely clear to me on why we need to copy. To be clear, I'm not against it but I will like to under why. We have similar implementations that we don't copy so there seems to be a miss alignment. Ex: `consensus-types/blocks/execution.go`

File: math/math_helper.go [hunk 13]
```diff
@@ -213,14 +213,21 @@ func AddInt(i ...int) (int, error) {
 	return sum, nil
 }
 
+// Wei is the smallest unit of Ether, and sent in the payload value of the execution engine API call.
+// represented as a pointer to a bigInt.
```
Comment (rkapka) [inline math/math_helper.go:217]:
```suggestion
// Wei is the smallest unit of Ether, represented as a pointer to a bigInt.
```
I don't think a type definition should specify a particular use case, although it's a nitpick really.

**Review by rkapka — APPROVED** [review]

**Pull request merged**